src/connector/src/source/nats/source/reader.rs

```diff
@@ -0,0 +1,82 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::Result;
+use async_trait::async_trait;
+use futures::{StreamExt, TryStreamExt};
+use futures_async_stream::try_stream;
+
+use crate::impl_common_split_reader_logic;
+use crate::parser::ParserConfig;
+use crate::source::nats::NatsProperties;
+use crate::source::{
+    BoxSourceWithStateStream, Column, SourceContextRef, SourceMessage, SplitImpl, SplitReader,
+};
+
+impl_common_split_reader_logic!(NatsSplitReader, NatsProperties);
+
+pub struct NatsSplitReader {
+    subscriber: async_nats::Subscriber,
+    properties: NatsProperties,
+    parser_config: ParserConfig,
+    source_ctx: SourceContextRef,
+}
+
+#[async_trait]
+impl SplitReader for NatsSplitReader {
+    type Properties = NatsProperties;
+
+    async fn new(
+        properties: NatsProperties,
+        splits: Vec<SplitImpl>,
+        parser_config: ParserConfig,
+        source_ctx: SourceContextRef,
+        _columns: Option<Vec<Column>>,
+    ) -> Result<Self> {
+        // TODO: to simplify the logic, return 1 split for first version
+        assert!(splits.len() == 1);
+        let subscriber = properties.common.build_subscriber().await?;
+        Ok(Self {
+            subscriber,
+            properties,
+            parser_config,
+            source_ctx,
+        })
+    }
+
+    fn into_stream(self) -> BoxSourceWithStateStream {
+        self.into_chunk_stream()
+    }
+}
+
+impl NatsSplitReader {
+    #[try_stream(boxed, ok = Vec<SourceMessage>, error = anyhow::Error)]
+    async fn into_data_stream(self) {
+        // TODO: how to deal with msg by batch
+        let mut subscriber = self.subscriber;
+        while let Some(message) = subscriber.next().await {
+            yield vec![SourceMessage::from_nats_message(message)]
```

Comment:
emit one message at a time may is not efficient.

Reply:
Yeah, agree with you. I find other conenctor will deal with the message as a batch. For nats, it use subscriber to get the message. Every time we call it, it will get one message. I try to deal with it manually, for example get N messages and yeild them in the vector. But I find if number of message is not enough N, it will stuck at there and waiting for incoming message.

Reply:
I find kafka use `read_chunks`. Then use it here. Do we need more optimization?

Reply:
> Do we need more optimization?

I think no need for that. Btw, can we get some number of the throughput?